fendermint/vm/topdown/src/sync.rs

```diff
@@ -0,0 +1,304 @@
+
+        let mut interval = tokio::time::interval(Duration::from_secs(config.polling_interval_secs));
+
+        tokio::spawn(async move {
+            loop {
+                interval.tick().await;
+
+                if let Err(e) = sync_with_parent(&config, &parent_client, &provider, &query).await {
+                    tracing::error!("sync with parent encountered error: {e}");
+                }
+            }
+        });
+    }
+}
+
+/// Syncing with parent with the below steps:
+/// 1. Get the latest height in cache or latest height committed increment by 1 as the
+///    starting height
+/// 2. Get the latest chain head height deduct away N blocks as the ending height
+/// 3. Fetches the data between starting and ending height
+/// 4. Update the data into cache
+async fn sync_with_parent<T: ParentFinalityStateQuery + Send + Sync + 'static>(
+    config: &Config,
+    parent_proxy: &Arc<IPCProviderProxy>,
+    provider: &Arc<Toggle<CachedFinalityProvider<IPCProviderProxy>>>,
+    query: &Arc<T>,
+) -> anyhow::Result<()> {
+    let last_recorded_height = if let Some(h) = last_recorded_height(provider).await? {
+        h
+    } else {
+        // cannot get starting recorded height, we just wait for the next loop execution
+        return Ok(());
+    };
+
+    let parent_chain_head_height = parent_proxy
+        .get_chain_head_height()
+        .await
+        .context("cannot fetch parent chain head")?;
+    // sanity check
+    if parent_chain_head_height < config.chain_head_delay {
+        tracing::debug!("latest height not more than the chain head delay");
+        return Ok(());
+    }
+
+    let ending_height = parent_chain_head_height - config.chain_head_delay;
+
+    tracing::debug!(
+        "last recorded height: {}, parent chain head: {}, ending_height: {}",
+        last_recorded_height,
+        parent_chain_head_height,
+        ending_height
+    );
+
+    if last_recorded_height == ending_height {
+        tracing::debug!(
+            "the parent has yet to produce a new block, stops at height: {last_recorded_height}"
+        );
+        return Ok(());
+    }
+
+    // we are going backwards in terms of block height, the latest block height is lower
+    // than our previously fetched head. It could be a chain reorg. We clear all the cache
+    // in `provider` and start from scratch
+    if last_recorded_height > ending_height {
+        return reset_cache(parent_proxy, provider, query).await;
+    }
+
+    // we also check if the starting height is the one we keep in the provider, if not
+    // this also means that there's been a reorg.
+    let last_committed_block = atomically(|| {
+        let b = if let Some(f) = provider.last_committed_finality()? {
+            Some(f.block_hash)
+        } else {
+            None
+        };
+        Ok(b)
+    })
+    .await;
+    if let Some(blk) = last_committed_block {
+        if blk
+            != parent_proxy
+                .get_block_hash(last_recorded_height)
```

Comment:
I think you are comparing different things here. 

`parent_proxy.get_block_hash(last_recorded_height)` is the correct thing to fetch, but it shouldn't be compared to the `provider.last_committed_finality()` because that one only moves when something is marked as final, while the other one moves every time we poll and there are new blocks. They do not necessarily point at the same thing. 

Somehow you would have to get from the cache the hash of the `last_recorded_height` and validate that it's still valid, and can be extended with new blocks.

Reply:
You can see here that they aren't the same: https://github.com/consensus-shipyard/fendermint/pull/308/files#diff-2d785f545d077d6b6d959f0ccac11391a65407b67a677c47ff6ba83192edf7f0R254-R257

Reply:
I see. I will revert the change and let @cryptoAtwill handle this one (I feel he has the full context and I am reading something wrong).